Automattic/jetpack
# Add a confirmation modal window when deactivating Jetpack

Fixes #15858 

This PR is part of the efforts to better handle disconnection (see p1HpG7-8VJ-p2) and adds a confirmation modal window when the user tries to Deactivate Jetpack while other plugins are relying on the Jetpack connection.

### Background

Now that the Connection package is being used by other plugins, we want to avoid the situation where the user break other plugins when they deactivate Jetpack.

Since Jetpack clear the connection tokens when deactivated, this would break other active plugins relying on the connection. We want to warn users about this.

We've already added this warning to the "Manage site connection" dialog in the Jetpack Dashboard (see #15456) and, after some work with the design tem (p6TEKc-3FC-p2), we made this warning more prominent (#15818).

We agreed that the same warning should be shown at deactivation, and the original idea was to reuse the same modal. After working on #14965 we realized it would not be possible in the short run, and we have to rethink the whole disconnection flow. 

But since this warning is important, especially now with WCPayments being launched, we decided to go for a new modal and worked on this PR, building it from scratch and only displaying the list of active plugins (discarding all the other contents from the disconnection modal).

### Changes proposed in this Pull Request:

* In the plugins screen, if the user is connected and there is one or more other plugins using Jetpack connection, it:
 * Changes the label of the Deactivate link to "Disconnect and Deactivate"
 * When the user clicks it, it opens a modal window with a confirmation dialog displaying all other plugins that are using the connection and warning the user that they will lose connection if Jetpack is deactivated.

### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* It adds a new confirmation step when the user is deactivating Jetpack (and has other plugins using connection)

### Challenges

The main challenge here is to make this modal look like the React `Modal` component used in the Jetpack Dashboard.

We are using Thickbox, and it will auto calculate the modal height when the window is opened and also at every screen resize, so it's hard to have full control of the height of the elements. 

Since the list of plugins will not grow in the near future, I favored a small window by default. It won't grow in height, so when the list of plugin grows, it will add a scrollbar (see screenshots).

#### Known issue

Thickbox changes the modal position at every window resize. I manage to override this in the first load, so I vertical centralized the modal, but if you resize the window, it will go back to the top. Also, I was not able to horizontal align it because Thickbox is doing this in a weird way, using a negative value for `margin-left` hard to calculate.

### Testing instructions:

* Set up a site using this branch and connect Jetpack
* Visit Plugins page on the Admin dashboard
* Check and see that the "Deactivate" link for the Jetpack plugin looks the same and nothing changed
* Create a file `wp-content/plugins/jetpack-connection-test-1.php` with following content:
```
<?php
/** Plugin Name: Jetpack Connection Test Plugin (1) */
add_action( 'plugins_loaded', function() {
	( new Automattic\Jetpack\Config() )->ensure( 'connection', array( 'slug' => 'jetpack-connection-test-1', 'name' => 'Jetpack Connection Test Plugin (1)' ) );
}, 1 );
```
* Activate this new plugin
* Now check that the label to deactivate Jetpack says "Disconnect and Deactivate"
* Click "Disconnect and Deactivate" and check that it opens a modal looking like the screenshots below
* Click Cancel and the close button at the top and confirm it closes the window
* Click the red "Disconnect and Deactivate" button and confirm it deactivates Jetpack

#### Layout tests

* Repeat the tests above resizing the window in different sizes
* Repeat the tests emulating mobile resolutions
* Repeat the tests in different browsers
* Tweak the HTML and make the list of plugins grow until it adds a vertical scrollbar to the modal.
 * Check again in all screen sizes with this extended list

#### Screenshots

Where you click to open the modal:
![Captura de tela de 2020-05-19 19-11-17](https://user-images.githubusercontent.com/971483/82385934-a1e8e780-9a09-11ea-9ce3-187dcc4a51f6.png)

Modal window opened:

![Captura de tela de 2020-05-20 17-48-14](https://user-images.githubusercontent.com/971483/82496952-e2a53700-9ac3-11ea-97fd-818843781327.png)

Big list with scroll:

![Captura de tela de 2020-05-20 17-49-11](https://user-images.githubusercontent.com/971483/82496950-e20ca080-9ac3-11ea-9f48-abb5c093e9e8.png)

![Captura de tela de 2020-05-20 17-49-24](https://user-images.githubusercontent.com/971483/82496946-e0db7380-9ac3-11ea-8d50-be800c8b49dc.png)

Mobile:

![photo_2020-05-21_12-06-51](https://user-images.githubusercontent.com/971483/82573273-f4d2b400-9b5b-11ea-840c-d249ee83cdf5.jpg)

### Proposed changelog entry for your changes:

* Adds a confirmation dialog upon deactivation when there are other plugins using connection
